engine/runtime/src/main/java/org/enso/interpreter/node/expression/builtin/meta/LazyAtomInstanceNode.java

```diff
@@ -0,0 +1,163 @@
+package org.enso.interpreter.node.expression.builtin.meta;
+
+import com.oracle.truffle.api.CompilerDirectives;
+import com.oracle.truffle.api.dsl.Cached;
+import com.oracle.truffle.api.dsl.Specialization;
+import com.oracle.truffle.api.frame.VirtualFrame;
+import com.oracle.truffle.api.interop.ArityException;
+import com.oracle.truffle.api.interop.InteropLibrary;
+import com.oracle.truffle.api.interop.TruffleObject;
+import com.oracle.truffle.api.interop.UnsupportedMessageException;
+import com.oracle.truffle.api.interop.UnsupportedTypeException;
+import com.oracle.truffle.api.library.CachedLibrary;
+import com.oracle.truffle.api.library.ExportLibrary;
+import com.oracle.truffle.api.library.ExportMessage;
+import com.oracle.truffle.api.nodes.Node;
+import com.oracle.truffle.api.nodes.RootNode;
+import org.enso.interpreter.dsl.BuiltinMethod;
+import org.enso.interpreter.runtime.EnsoContext;
+import org.enso.interpreter.runtime.callable.argument.ArgumentDefinition;
+import org.enso.interpreter.runtime.callable.argument.CallArgumentInfo;
+import org.enso.interpreter.runtime.callable.atom.Atom;
+import org.enso.interpreter.runtime.callable.function.Function;
+import org.enso.interpreter.runtime.callable.function.FunctionSchema;
+import org.enso.interpreter.runtime.data.Array;
+import org.enso.interpreter.runtime.data.Vector;
+import org.enso.interpreter.runtime.error.PanicException;
+
+@BuiltinMethod(
+    type = "Meta",
+    name = "lazy_atom",
+    description = "Creates a new atom with given constructor and fields.",
+    autoRegister = false)
+public abstract class LazyAtomInstanceNode extends Node {
+
+  static LazyAtomInstanceNode build() {
+    return LazyAtomInstanceNodeGen.create();
+  }
+
+  abstract Vector execute(Object factory);
+
+  @Specialization(guards = "iop.isExecutable(factory)")
+  Vector doExecute(
+    Object factory,
+    @CachedLibrary(limit="3") InteropLibrary iop,
```

Comment:
I'd rather use `InvokeCallableNode` here, not `InteropLibrary` – the former is better with things like state etc. and may be faster with enso lambdas in general.

Reply:
Done in [b383ace](https://github.com/enso-org/enso/pull/4023/commits/b383acebe92a9ab7f2476ccc45ed9af64c134f0d) - now the `ListBenchmarks` is twice fast than with `InteropLibrary`, but it still needs some love.